src/cache/cache_las.c

```diff
@@ -1086,8 +1090,14 @@ __wt_las_sweep(WT_SESSION_IMPL *session)
      */
     __wt_las_cursor(session, &cursor, &session_flags);
     WT_ASSERT(session, cursor->session == &session->iface);
-    WT_ERR(__wt_txn_begin(session, NULL));
     __las_set_isolation(session, &saved_isolation);
+
```

Comment:
@kommiharibabu the lookaside cursors need to use `read uncommitted`. I think of it like this: we are writing content into the lookaside file that contains the information required to decide whether something is visible. Any lookaside cursor should return all available data, then the thing digesting the information uses the timestamp/txnID information stored into the lookaside file to make decisions about visibility.

Reply:
@agorrod I agree with you regarding the LAS read with `read uncommitted`. It works even for the `sweep server` also in the `develop` because from the LAS records itself it is possible to identify whether the disk image is written the newest version or not? If the disk image is visible to everyone, the entire contents of the LAS records related to the key can be removed.

But with `relevant history` there is no way to identify whether the disk image record is the newest page record or not? So we cannot delete the entire key block without checking each individual record. And with `modify` records, we should not remove the complete update records until there is a complete update record is found for the same key. Because of this reason, the `prev` record can only be deleted once the `next` record is not of modify type or there is a complete record.

With `read uncommitted` isolation, there is a chance the LAS prev operation may get a completely new record that is inserted newly than the record that we visited previously. Because of this reason, I do the `snapshot` isolation to get the same record that it visited last time.

Please let me know your views on this problem, is it not possible at all? or any better solution?

Reply:
@keithbostic, I don't think the problem can occur for the `same key` values, as there is no way that a new record that can be inserted in-between `same key` data. The problem can happen only in the boundary cases of moving from one key to another.

Let us take this example, Following are the LAS records. 
`key1 (10), timestamp (10), txnid (10)` -> `key2(20), timestamp(10), txnid(10)`

The sweep server identifies that `key1(10), timestamp(10), txnid(10)` record can be removed from LAS as it is obsolete, but to do that, it must verify whether the next record is not a modify operation. So the scan proceeds to the next record `key2(20), timestamp(10), txnid(10)` and decides that the key doesn't match, so the previous record can be deleted. By that time, it is possible that a new record for the `key1(10)` can get inserted with data as `key1(10), timestamp(40), txnid (40)`.

In the above scenario, if the sweep server continues to remove the `prev` record, it will remove the newly inserted record `key1(10), timestamp(40), txnid(40)`.

With the `isolation` snapshot, I think that this problem doesn't occur. I am not sure exactly how internally it works. Am I correct?

Another alternative to this problem is doing a `backward` scan of the LAS records and remove the record whenever it found as obsolete. There is no need to worry about `isolation` and `modify` record handling. opinions?

Reply:
I believe your analysis is correct.

I also agree with your comment to @tetsuo-cpp that we are better served with a cursor-next approach to sweeping the LAS file.

An alternative I think might work (while allowing a read-uncommitted configuration), is to used the saved key/timestamp/txnid triplet to walk backward through the LAS file, discarding the matches. That is, cursor.prev backward until the number of matching saved-key modify records have been deleted, and then resume forward traversal. There's a count of modify records to be deleted, so the cursor.prev movement is bounded. We've added some get-key/get-value comparisons, but that's about all.

I agree this is more complicated in one dimension, but it would simplify the existing (fairly complex) loop logic, and also allow us to continue to use a read-uncommitted configuration.

I will leave it up to you, Alex and Vamsi which approach is best.

Reply:
@keithbostic, @kommiharibabu moved this conversation into JIRA.